pymc/model.py

```diff
@@ -2035,10 +2035,49 @@ def Deterministic(name, var, model=None, dims=None):
 def Potential(name, var, model=None):
     """Add an arbitrary factor potential to the model likelihood
 
+      The Potential function is used to add arbitrary factors (such as constraints or other likelihood components) to adjust the probability density of the model.
+
+      Have a look at the following example:
+
+     .. code:: python
+
+        with pm.Model() as model:
+          x = pm.Normal('x', mu=0, sigma=1)
+          y = pm.Normal('y', mu=x, sigma=1, observed=data)
+          constraint = x>= 0
+          potential = pm.Potential('x_constraint', -pm.math.log(pm.math.switch(constraint, 1, 1e-30))
```

Comment:
I think the absolute constraint with (1 vs 0) may be more educative than (1 vs 1e-30). I would perhaps show that and mention an alternative something like (1 vs 0.5) would provide a soft constraint instead of a hard one.

I find the 1e-30 odd because it behaves like 0 in practice. The sampler will never accept a negative value, or perhaps once in a lifetime, which is an odd modelling choice.

We should change the mu of x to be 0.1 perhaps, so that the sampler doesn't start at an invalid point.

Reply:
I understand that the absolute constraint with (1 vs 0) might be more educative than (1 vs 1e-30), but the problem with it is that it produces error when trying to sample the model because the logarithm of 0 is not defined. If having a working model that can be sampled without encountering errors along with explaining the use of Potential function is not what is required with this code block, a constraint of 1 vs 0 absolutely makes more sense. So to reconfirm, should I change the constraint?

Soft constraints also are very helpful. Should I make a  separate code-block for it or a commented line in the original code-block with an explanation?

Reply:
Logarithm of zero is well defined (at least in PyMC land) and leads to -inf logp which is exactly what you want to represent an impossible event in log scale. We use -inf explicitly in most of the logp definitions with constrained support.

About the soft constrained, whichever way you think is better. I have no preference.

Reply:
When I try to sample using potential = pm.Potential('x_constraint',-pm.math.log(pm.math.switch(constraint, 1, 0.0))) I face this error:

AttributeError                            Traceback (most recent call last)
[<ipython-input-31-e645e4d4afc7>](https://localhost:8080/#) in <module>
     11   potential = pm.Potential('x_constraint',-pm.math.log(pm.math.switch(constraint, 1, 0)))
     12 
---> 13   trace = pm.sample(1000,tune=1000)

12 frames
[/usr/local/lib/python3.8/dist-packages/aesara/tensor/elemwise.py](https://localhost:8080/#) in transform(r)
    600         def transform(r):
    601             # From a graph of ScalarOps, make a graph of Broadcast ops.
--> 602             if isinstance(r.type, (NullType, DisconnectedType)):
    603                 return r
    604             if r in scalar_inputs:

AttributeError: 'float' object has no attribute 'type'

On defining the Potential function as potential = pm.Potential('x_constraint',-pm.math.log(pm.math.switch(constraint, 1.0, 0.0))) I get this:

Interrupted
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-34-ed86f6db8804>](https://localhost:8080/#) in <module>
     11   potential = pm.Potential('x_constraint',-pm.math.log(pm.math.switch(constraint, 1.0, 0.0)))
     12 
---> 13   trace = pm.sample(1000,tune=1000)

8 frames
[/usr/local/lib/python3.8/dist-packages/pymc/step_methods/hmc/base_hmc.py](https://localhost:8080/#) in astep(self, q0)
    166             message_energy = (
    167                 "Bad initial energy, check any log probabilities that "
--> 168                 "are inf or -inf, nan or very small:\n{}".format(error_logp.to_string())
    169             )
    170             warning = SamplerWarning(

AttributeError: 'numpy.ndarray' object has no attribute 'to_string'

Reply:
For it to have a valid initial energy with the hard constraint you have to set the prior of x to not be centered at zero like `x = pm.Normal("x", mu=0.1, sigma=1)` or set an initial value like `x = pm.Normal("x", mu=0, sigma=1, initval=0.1)`

Reply:
The negative sign was the problem. I think this is because if we take the negative sign, the Potential function takes in inf as argument instead of -inf.  From my understanding, the potential term should be somewhere close to 0 when the constraint is satisfied and very large when the constraint is not satisfied. What I thought was that both inf and -inf can work as this large value. But when Potential value becomes inf,  for some reason I face the error "'numpy.ndarray' object has no attribute 'to_string'". So, is it wrong if the Potential value becomes inf?

Reply:
The negative sign was the problem. I think this is because if we take the negative sign, the Potential function returns in inf instead of -inf. From my understanding, the potential term should be somewhere close to 0 when the constraint is satisfied and very large when the constraint is not satisfied. What I thought was that both inf and -inf can work as this large value. But when Potential value becomes inf, for some reason I face the error "'numpy.ndarray' object has no attribute 'to_string'". So, is it wrong if the Potential value becomes inf?

Reply:
The potential value is just added to the model logp, so it should be -inf (or very negative) when a constraint is violated (so that those draws are rejected. 0 won't have any effect and positive values will make the proposals more likely to be accepted

Reply:
Alright👍 
I shall include this in the docstring too.
Thanks for the help!